### PR TITLE
Support opt-in Non-Selector SyncSet first applied metrics

### PR DESCRIPTION
**Pull request by akhil-rane**
This PR adds support to group non-selector syncsets using 'hive.openshift.io/syncset-metrics-group' 
annotation and report first applied metrics for each group.

jira: https://issues.redhat.com/browse/CO-1217

### CODE REVIEW

**Comment by akhil-rane** [comment]
/assign @suhanime 
/cc @dgoodwin 
  
I am yet to test it properly but look out for obvious issues in the code.

File: pkg/constants/constants.go [hunk 1]
```diff
@@ -301,6 +301,10 @@ const (
 	// ReconcileIDLen is the length of the random strings we generate for contextual loggers in controller
 	// Reconcile functions.
 	ReconcileIDLen = 8
+
+	// syncSetMetricsGroupAnnotation can be applied to SyncSets to make them part of a
```
Comment (suhanime) [inline pkg/constants/constants.go:305]:
```suggestion
	// SyncSetMetricsGroupAnnotation can be applied to non-selector SyncSets to make them part of a
```

File: pkg/controller/controlplanecerts/controlplanecerts_controller.go [hunk 2]
```diff
@@ -263,8 +263,9 @@ func (r *ReconcileControlPlaneCerts) generateControlPlaneCertsSyncSet(cd *hivev1
 	cdLog.Debug("generating syncset for control plane secrets")
 	syncSet := &hivev1.SyncSet{
 		ObjectMeta: metav1.ObjectMeta{
-			Name:      GenerateControlPlaneCertsSyncSetName(cd.Name),
-			Namespace: cd.Namespace,
+			Name:        GenerateControlPlaneCertsSyncSetName(cd.Name),
+			Namespace:   cd.Namespace,
+			Annotations: map[string]string{constants.SyncSetMetricsGroupAnnotation: "cp-certs"},
```
Comment (suhanime) [inline pkg/controller/controlplanecerts/controlplanecerts_controller.go:268]: changes in controllers can go in as a separate commit to document the label already present on the generated syncsets

File: pkg/controller/clustersync/clustersync_controller.go [hunk 3]
```diff
@@ -541,6 +550,10 @@ func (r *ReconcileClusterSync) applySyncSets(
 				// or SyncSet creationTimestamp.
 				logger.WithField("applyTime", applyTime).Debug("observed first successful apply of SyncSet for cluster")
 				metricTimeToApplySyncSet.Observe(applyTime)
+
+				if syncSetGroup, ok := syncSet.AsMetaObject().GetAnnotations()[constants.SyncSetMetricsGroupAnnotation]; ok {
+					metricTimeToApplyGroupedSyncSet.WithLabelValues(syncSetGroup).Observe(applyTime)
+				}
```
[the remark this answers is not in the document]
Comment (dgoodwin) [reply pkg/controller/clustersync/clustersync_controller.go:556]: My vote would be to keep them reported in metricTimeToApplySyncSet, that should be over all SyncSets and it would get hard to explain if it were just those without an explicit grouping.

I really like your idea of combining them to avoid repetition. Lets add the group label to metricTimeToApplySyncSet. I think prometheus would require us to set some value for those without a group, maybe "none" or something like that?

[the remark this answers is not in the document]
Comment (suhanime) [reply pkg/controller/clustersync/clustersync_controller.go:556]: Yes - grouping sounds the most ideal. We could label the syncsets without a metrics label as "none".
From a statistical perspective, it makes sense to keep them clubbed under 1 graph - so we can simultaneously see the stats for the grouped SyncSets as well as apply aggregation to derive the results for all SyncSets.
@akhil-rane does this seem okay to you?

[the remark this answers is not in the document]
Comment (akhil-rane) [reply pkg/controller/clustersync/clustersync_controller.go:556]: looks good to me

**Comment by akhil-rane** [comment]
@dgoodwin @suhanime here is what I get in prometheus with and without grouping annotation. I suppose this is what we are looking for

`[akhilrane@localhost hive]$ curl http://localhost:2112/metrics | grep hive_syncset_apply_duration_seconds
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0# HELP hive_syncset_apply_duration_seconds Time to first successfully apply syncset to a cluster
# TYPE hive_syncset_apply_duration_seconds histogram
hive_syncset_apply_duration_seconds_bucket{group="none",le="5"} 1
hive_syncset_apply_duration_seconds_bucket{group="none",le="10"} 1
hive_syncset_apply_duration_seconds_bucket{group="none",le="30"} 1
hive_syncset_apply_duration_seconds_bucket{group="none",le="60"} 1
hive_syncset_apply_duration_seconds_bucket{group="none",le="300"} 1
hive_syncset_apply_duration_seconds_bucket{group="none",le="600"} 1
hive_syncset_apply_duration_seconds_bucket{group="none",le="1800"} 1
hive_syncset_apply_duration_seconds_bucket{group="none",le="3600"} 1
hive_syncset_apply_duration_seconds_bucket{group="none",le="+Inf"} 1
hive_syncset_apply_duration_seconds_sum{group="none"} 1.7169265660000002
hive_syncset_apply_duration_seconds_count{group="none"} 1
hive_syncset_apply_duration_seconds_bucket{group="test-group",le="5"} 1
hive_syncset_apply_duration_seconds_bucket{group="test-group",le="10"} 1
hive_syncset_apply_duration_seconds_bucket{group="test-group",le="30"} 1
hive_syncset_apply_duration_seconds_bucket{group="test-group",le="60"} 1
hive_syncset_apply_duration_seconds_bucket{group="test-group",le="300"} 1
hive_syncset_apply_duration_seconds_bucket{group="test-group",le="600"} 1
hive_syncset_apply_duration_seconds_bucket{group="test-group",le="1800"} 1
hive_syncset_apply_duration_seconds_bucket{group="test-group",le="3600"} 1
hive_syncset_apply_duration_seconds_bucket{group="test-group",le="+Inf"} 1
hive_syncset_apply_duration_seconds_sum{group="test-group"} 0.902421274
hive_syncset_apply_duration_seconds_count{group="test-group"} 1
100  204k    0  204k    0     0   360k      0 --:--:-- --:--:-- --:--:--  360k`

**Comment by suhanime** [comment]
> @dgoodwin @suhanime here is what I get in prometheus with and without grouping annotation. I suppose this is what we are looking for
> 
> ```
> [akhilrane@localhost hive]$ curl http://localhost:2112/metrics | grep hive_syncset_apply_duration_seconds
>   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
>                                  Dload  Upload   Total   Spent    Left  Speed
>   0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0# HELP hive_syncset_apply_duration_seconds Time to first successfully apply syncset to a cluster
> # TYPE hive_syncset_apply_duration_seconds histogram
> hive_syncset_apply_duration_seconds_bucket{group="none",le="5"} 1
> hive_syncset_apply_duration_seconds_bucket{group="none",le="10"} 1
> hive_syncset_apply_duration_seconds_bucket{group="none",le="30"} 1
> hive_syncset_apply_duration_seconds_bucket{group="none",le="60"} 1
> hive_syncset_apply_duration_seconds_bucket{group="none",le="300"} 1
> hive_syncset_apply_duration_seconds_bucket{group="none",le="600"} 1
> hive_syncset_apply_duration_seconds_bucket{group="none",le="1800"} 1
> hive_syncset_apply_duration_seconds_bucket{group="none",le="3600"} 1
> hive_syncset_apply_duration_seconds_bucket{group="none",le="+Inf"} 1
> hive_syncset_apply_duration_seconds_sum{group="none"} 1.7169265660000002
> hive_syncset_apply_duration_seconds_count{group="none"} 1
> hive_syncset_apply_duration_seconds_bucket{group="test-group",le="5"} 1
> hive_syncset_apply_duration_seconds_bucket{group="test-group",le="10"} 1
> hive_syncset_apply_duration_seconds_bucket{group="test-group",le="30"} 1
> hive_syncset_apply_duration_seconds_bucket{group="test-group",le="60"} 1
> hive_syncset_apply_duration_seconds_bucket{group="test-group",le="300"} 1
> hive_syncset_apply_duration_seconds_bucket{group="test-group",le="600"} 1
> hive_syncset_apply_duration_seconds_bucket{group="test-group",le="1800"} 1
> hive_syncset_apply_duration_seconds_bucket{group="test-group",le="3600"} 1
> hive_syncset_apply_duration_seconds_bucket{group="test-group",le="+Inf"} 1
> hive_syncset_apply_duration_seconds_sum{group="test-group"} 0.902421274
> hive_syncset_apply_duration_seconds_count{group="test-group"} 1
> 100  204k    0  204k    0     0   360k      0 --:--:-- --:--:-- --:--:--  360k
> ```

Looks good.
And in grafana, if we aggregate without group, it will give us stats for all the SyncSets :)

File: pkg/constants/constants.go (hunk 1 above)
Comment (akhil-rane) [reply pkg/constants/constants.go:305]: This sounds better.Thanks :)

File: pkg/controller/clustersync/clustersync_controller.go [hunk 4]
```diff
@@ -540,7 +541,11 @@ func (r *ReconcileClusterSync) applySyncSets(
 				// For non-selector SyncSets we have a more accurate startTime, either ClusterDeployment installedTimestamp
 				// or SyncSet creationTimestamp.
 				logger.WithField("applyTime", applyTime).Debug("observed first successful apply of SyncSet for cluster")
-				metricTimeToApplySyncSet.Observe(applyTime)
+				if syncSetGroup, ok := syncSet.AsMetaObject().GetAnnotations()[constants.SyncSetMetricsGroupAnnotation]; ok {
+					metricTimeToApplySyncSet.WithLabelValues(syncSetGroup).Observe(applyTime)
+				} else {
+					metricTimeToApplySyncSet.WithLabelValues("non-grouped").Observe(applyTime)
```
Comment (akhil-rane) [inline pkg/controller/clustersync/clustersync_controller.go:547]: If labels are used in graph index does ''non-grouped' sound better than "none"? I am not sure here.

**Comment by akhil-rane** [comment]
PR updated
